django/db/models/fields/related_descriptors.py

```diff
@@ -1130,6 +1131,39 @@ def get_prefetch_queryset(self, instances, queryset=None):
                 False,
             )
 
+        @property
+        def constrained_target(self):
+            # If the through relation's target field's foreign integrity is
+            # enforced, the query can be performed solely against the through
+            # table as the INNER JOIN'ing against target table is unnecessary.
+            if not self.target_field.db_constraint:
+                return None
+            db = router.db_for_read(self.through, instance=self.instance)
+            if not connections[db].features.supports_foreign_keys:
+                return None
+            hints = {'instance': self.instance}
+            manager = self.through._base_manager.db_manager(db, hints=hints)
+            filters = {self.source_field_name: self.instance.pk}
+            # Nullable target rows must be excluded as well as they would have
+            # been filtered out from an INNER JOIN.
+            if self.target_field.null:
+                filters['%s__isnull' % self.target_field_name] = False
+            return manager.filter(**filters)
+
+        def exists(self):
+            constrained_target = self.constrained_target
+            if constrained_target and ManyRelatedManager.__class__ == models.Manager:
```

Comment:
This critera will always be false as `ManyRelatedManager` is a subclass of `models.Manager`. Pretty sure you want to check against `superclass` instead?

Reply:
Wow another bug with 1299421cadc4fcf63585f2f88337078e43e660e0 is that the checking for `constrained_target` instead of `constrained_target is not None` triggers `QuerySet.__bool__` which causes an extra query!

The only reason the tests pass is because the first query that is checked for `JOIN` is the `SELECT` that doing `bool(constrained_target)` generates.

Reply:
Thanks for your patience, @charettes, I'm currently trying to add coverage for the prefetched case in comment 18 on ticket https://github.com/django/django/pull/10366.

Will the check, `and self.get_prefetch_queryset(self, self.instances)._prefetch_done is False:` when appended to line 1155 correctly determine whether or not the items have been prefetched? Thanks for your help.

Reply:
> @aprilschen you have a few linting errors to address and I left some extra comment to properly tests these changes. You'll need to make sure to actually add tests to cover the case where a custom manager class is used that should disable the optimization.
> 
> Also, prefixing your PR description with `Fixed #29725 -- ...` would make it easier to link it back to Trac.

Oops, looks like I still have more linting errors to address after this new PR. I will be doing so in the next commit!